File: question_9.py
from itertools import combinations, permutations
import sys
from typing import List, Tuple

class ParcelRouter:
    def __init__(self, office_a: Tuple[int, int], office_b: Tuple[int, int], collection_points: List[Tuple[int, int]]):
        self.office_a = office_a
        self.office_b = office_b
        self.collection_points = collection_points
        self.n = len(collection_points)

    def calculate_distance(self, point1: Tuple[int, int], point2: Tuple[int, int]) -> int:
        """Calculate Manhattan distance between two points"""
        return abs(point1[0] - point2[0]) + abs(point1[1] - point2[1])

    def calculate_route_fuel(self, office: Tuple[int, int], points: List[Tuple[int, int]]) -> int:
        """Calculate total fuel consumption for a single truck route"""
        if not points:
            return 0

        fuel = 0
        current_pos = office
        parcels = 0

        # Calculate fuel for each segment of the route
        for point in points:
            distance = self.calculate_distance(current_pos, point)
            fuel += distance * (1 + parcels)  # Fuel consumption formula
            current_pos = point
            parcels += 1  # Add one parcel at each collection point

        # Return trip to office
        distance = self.calculate_distance(current_pos, office)
        fuel += distance * (1 + parcels)

        return fuel

    def find_minimum_fuel_route(self) -> int:
        """Find the minimum total fuel consumption for both trucks"""
        min_total_fuel = sys.maxsize

        # Try all possible combinations of points between the two trucks
        for i in range(self.n + 1):  # Number of points for truck A (including 0)
            for combo in combinations(range(self.n), i):
                # Points for truck A
                points_a = [self.collection_points[idx] for idx in combo]
                # Remaining points go to truck B
                points_b = [self.collection_points[idx] for idx in range(self.n) if idx not in combo]

                # Try all possible orderings for truck A's points
                min_fuel_a = sys.maxsize if points_a else 0
                min_fuel_b = sys.maxsize if points_b else 0

                # Calculate minimum fuel for truck A's route
                if points_a:
                    for route in permutations(points_a):
                        fuel = self.calculate_route_fuel(self.office_a, route)
                        min_fuel_a = min(min_fuel_a, fuel)

                # Calculate minimum fuel for truck B's route
                if points_b:
                    for route in permutations(points_b):
                        fuel = self.calculate_route_fuel(self.office_b, route)
                        min_fuel_b = min(min_fuel_b, fuel)

                total_fuel = min_fuel_a + min_fuel_b
                min_total_fuel = min(min_total_fuel, total_fuel)

        return min_total_fuel

def main():
    # Read number of collection points
    N = int(input())

    # Read office locations
    x, y = map(int, input().split())
    office_a = (x, y)
    x, y = map(int, input().split())
    office_b = (x, y)

    # Read collection points
    collection_points = []
    for _ in range(N):
        x, y = map(int, input().split())
        collection_points.append((x, y))

    # Create router and find minimum fuel required
    router = ParcelRouter(office_a, office_b, collection_points)
    min_fuel = router.find_minimum_fuel_route()
    print(min_fuel)

if __name__ == "__main__":
    main()

File: question_9.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <climits>

using namespace std;

class ParcelRouter {
public:
    ParcelRouter(pair<int, int> office_a, pair<int, int> office_b, vector<pair<int, int>> collection_points)
        : office_a(office_a), office_b(office_b), collection_points(collection_points), n(collection_points.size()) {}

    int calculate_distance(pair<int, int> point1, pair<int, int> point2) {
        // Calculate Manhattan distance between two points
        return abs(point1.first - point2.first) + abs(point1.second - point2.second);
    }

    int calculate_route_fuel(pair<int, int> office, vector<pair<int, int>> points) {
        // Calculate total fuel consumption for a single truck route
        if (points.empty()) {
            return 0;
        }

        int fuel = 0;
        pair<int, int> current_pos = office;
        int parcels = 0;

        // Calculate fuel for each segment of the route
        for (auto point : points) {
            int distance = calculate_distance(current_pos, point);
            fuel += distance * (1 + parcels);  // Fuel consumption formula
            current_pos = point;
            parcels += 1;  // Add one parcel at each collection point
        }

        // Return trip to office
        int distance = calculate_distance(current_pos, office);
        fuel += distance * (1 + parcels);

        return fuel;
    }

    int find_minimum_fuel_route() {
        // Find the minimum total fuel consumption for both trucks
        int min_total_fuel = INT_MAX;

        // Try all possible combinations of points between the two trucks
        for (int i = 0; i <= n; ++i) {  // Number of points for truck A (including 0)
            vector<int> indices(n);
            fill(indices.begin(), indices.begin() + i, 1);
            do {
                // Points for truck A
                vector<pair<int, int>> points_a, points_b;
                for (int j = 0; j < n; ++j) {
                    if (indices[j]) {
                        points_a.push_back(collection_points[j]);
                    } else {
                        points_b.push_back(collection_points[j]);
                    }
                }

                // Try all possible orderings for truck A's points
                int min_fuel_a = points_a.empty() ? 0 : INT_MAX;
                int min_fuel_b = points_b.empty() ? 0 : INT_MAX;

                // Calculate minimum fuel for truck A's route
                if (!points_a.empty()) {
                    sort(points_a.begin(), points_a.end());
                    do {
                        int fuel = calculate_route_fuel(office_a, points_a);
                        min_fuel_a = min(min_fuel_a, fuel);
                    } while (next_permutation(points_a.begin(), points_a.end()));
                }

                // Calculate minimum fuel for truck B's route
                if (!points_b.empty()) {
                    sort(points_b.begin(), points_b.end());
                    do {
                        int fuel = calculate_route_fuel(office_b, points_b);
                        min_fuel_b = min(min_fuel_b, fuel);
                    } while (next_permutation(points_b.begin(), points_b.end()));
                }

                int total_fuel = min_fuel_a + min_fuel_b;
                min_total_fuel = min(min_total_fuel, total_fuel);

            } while (prev_permutation(indices.begin(), indices.end()));
        }

        return min_total_fuel;
    }

private:
    pair<int, int> office_a;
    pair<int, int> office_b;
    vector<pair<int, int>> collection_points;
    int n;
};

int main() {
    // Read number of collection points
    int N;
    cin >> N;

    // Read office locations
    int x, y;
    cin >> x >> y;
    pair<int, int> office_a = {x, y};
    cin >> x >> y;
    pair<int, int> office_b = {x, y};

    // Read collection points
    vector<pair<int, int>> collection_points(N);
    for (int i = 0; i < N; ++i) {
        cin >> x >> y;
        collection_points[i] = {x, y};
    }

    // Create router and find minimum fuel required
    ParcelRouter router(office_a, office_b, collection_points);
    int min_fuel = router.find_minimum_fuel_route();
    cout << min_fuel << endl;

    return 0;
}
